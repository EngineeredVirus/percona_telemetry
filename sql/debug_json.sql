--
-- json debugging
--

CREATE OR REPLACE FUNCTION read_json_file()
RETURNS text
AS $$
DECLARE
    file_path text;
    system_id text;
    pg_version text;
    file_content text;
BEGIN
    SELECT output_file_name INTO file_path FROM percona_telemetry_status();
    SELECT system_identifier::TEXT INTO system_id FROM pg_control_system();
    SELECT version() INTO pg_version;

    -- file_content := to_json('{' || pg_read_file(file_path) || '}');
    file_content := '{' || pg_read_file(file_path) || '}';

    RETURN file_content;
END;
$$ LANGUAGE plpgsql;

-- Let's sleep for a few seconds to ensure that leader has
-- generated the json file.
SELECT pg_sleep(3);

CREATE EXTENSION percona_telemetry;

SELECT 'db_instance_id' AS col FROM pg_control_system() WHERE '"' || system_identifier || '"' = (SELECT cast(read_json_file()::JSON->'db_instance_id' AS VARCHAR));
SELECT 'pillar_version' AS col WHERE '"' || current_setting('server_version') || '"' = (SELECT cast(read_json_file()::JSON->'pillar_version' AS VARCHAR));
SELECT 'databases_count' AS col FROM pg_database WHERE datallowconn = true HAVING '"' || count(*) || '"' = (SELECT cast(read_json_file()::JSON->'databases_count' AS VARCHAR));
SELECT 'settings' AS col FROM pg_settings WHERE name NOT LIKE 'plpgsql.%' HAVING count(*) = (SELECT json_array_length(read_json_file()::JSON->'settings'));

DROP EXTENSION percona_telemetry;
DROP FUNCTION read_json_file;