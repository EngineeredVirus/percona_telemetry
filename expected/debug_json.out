--
-- json debugging
--
CREATE OR REPLACE FUNCTION read_json_file()
RETURNS text
AS $$
DECLARE
    file_path text;
    system_id text;
    pg_version text;
    file_content text;
BEGIN
    SELECT output_file_name INTO file_path FROM percona_telemetry_status();
    SELECT system_identifier::TEXT INTO system_id FROM pg_control_system();
    SELECT version() INTO pg_version;

    file_content := pg_read_file(file_path);

    SELECT REPLACE(file_content, system_id, '<SYSTEM IDENTIFIER>') INTO file_content;
    SELECT REPLACE(file_content, pg_version, '<PG VERSION>') INTO file_content;

    -- If we have more than 6 consecutive digits, it's file size unlike to be an OID
    SELECT regexp_replace(file_content, 'archive_cleanup_command.*', 'archive_cleanup_command', 'g') INTO file_content;
    SELECT regexp_replace(file_content, '\d{2}:\d{2}:\d{2}.\d{6}', '<SERVER UPTIME>', 'g') INTO file_content;
    SELECT regexp_replace(file_content, '\d{2}\.\d{1,2}', '<PG VERSION>', 'g') INTO file_content;
    SELECT regexp_replace(file_content, '\d{6,}', '<DB SIZE>', 'g') INTO file_content;
    SELECT regexp_replace(file_content, '\d{1,}', '<DB OID>', 'g') INTO file_content;

    RETURN file_content;
END;
$$ LANGUAGE plpgsql;
-- Let's sleep for a few seconds to ensure that leader has
-- generated the json file.
SELECT pg_sleep(3);
 pg_sleep 
----------
 
(1 row)

CREATE EXTENSION percona_telemetry;
SELECT read_json_file();
                     read_json_file                      
---------------------------------------------------------
           "db_instance_id": "<SYSTEM IDENTIFIER>",     +
           "pillar_version": "<PG VERSION>",            +
           "uptime": "<SERVER UPTIME>",                 +
           "databases_count": "<DB OID>",               +
           "settings": [                                +
             {                                          +
               "key": "setting",                        +
               "value": [                               +
                 {                                      +
                   "key": "name",                       +
                   "value": "allow_in_place_tablespaces"+
                 },                                     +
                 {                                      +
                   "key": "unit",                       +
                   "value": "NULL"                      +
                 },                                     +
                 {                                      +
                   "key": "setting",                    +
                   "value": "off"                       +
                 },                                     +
                 {                                      +
                   "key": "reset_val",                  +
                   "value": "off"                       +
                 },                                     +
                 {                                      +
                   "key": "boot_val",                   +
                   "value": "off"                       +
                 }                                      +
               ]                                        +
             },                                         +
             {                                          +
               "key": "setting",                        +
               "value": [                               +
                 {                                      +
                   "key": "name",                       +
                   "value": "allow_system_table_mods"   +
                 },                                     +
                 {                                      +
                   "key": "unit",                       +
                   "value": "NULL"                      +
                 },                                     +
                 {                                      +
                   "key": "setting",                    +
                   "value": "off"                       +
                 },                                     +
                 {                                      +
                   "key": "reset_val",                  +
                   "value": "off"                       +
                 },                                     +
                 {                                      +
                   "key": "boot_val",                   +
                   "value": "off"                       +
                 }                                      +
               ]                                        +
             },                                         +
             {                                          +
               "key": "setting",                        +
               "value": [                               +
                 {                                      +
                   "key": "name",                       +
                   "value": "application_name"          +
                 },                                     +
                 {                                      +
                   "key": "unit",                       +
                   "value": "NULL"                      +
                 },                                     +
                 {                                      +
                   "key": "setting",                    +
                   "value": "NULL"                      +
                 },                                     +
                 {                                      +
                   "key": "reset_val",                  +
                   "value": "NULL"                      +
                 },                                     +
                 {                                      +
                   "key": "boot_val",                   +
                   "value": "NULL"                      +
                 }                                      +
               ]                                        +
             },                                         +
             {                                          +
               "key": "setting",                        +
               "value": [                               +
                 {                                      +
                   "key": "name",                       +
                   "value": "archive_cleanup_command
(1 row)

DROP EXTENSION percona_telemetry;
DROP FUNCTION read_json_file;
